R/functions_netCDF.R

```diff
@@ -456,42 +687,377 @@ create_netCDF_from_array_with_variables <- function(x, locations, grid,
+populate_netcdf_from_array <- function(file, data, var_names = NULL,
+                                       has_T_timeAxis, has_Z_verticalAxis,
+                                       isGridded = TRUE, grid = NULL, locations,
+                                       force_v4 = TRUE, verbose = FALSE) {
 
-  #--- add variables values
-  for (k in seq_len(nl)) {
-    # put values into full grid format
-    temp <- val_template
-    temp[val_ids] <- x[, k]
-    vals <- matrix(temp, ncol = var_chunksizes[2])
 
-    # write values to variable
-    try(ncdf4::ncvar_put(nc, varid = var_names[k], vals = vals,
-      start = var_start, count = var_chunksizes))
+  # ---------------------------------------------------------------------
+  # Set up and checks ---------------------------------------------------
+  # ---------------------------------------------------------------------
 
-    # add variable attributes
-    for (natt in ns_att_vars) {
-      ncdf4::ncatt_put(nc, varid = var_names[k], attname = natt,
-        attval = var_attributes[[natt]][k])
+  stopifnot(requireNamespace("ncdf4"))
+  if (force_v4) {
+    # avoid "_FillValue" error in older versions of `raster` package
+    stopifnot(utils::packageVersion("raster") >= "2.9.1")
+  }
+  
+  stopifnot(file.exists(file) || !missing(locations)) # file and locations need to exist
+
+  # open file, writeable --------------
+  nc <- ncdf4::nc_open(file, write = TRUE)
+  on.exit(ncdf4::nc_close(nc))
+  
+  # check  netcdf against data and inputs ---------------------------
+  nc_dims <-  attributes(nc$dim)$names #dims of netcdf
+  nn <- NCOL(data) # number of cols of data
+  data_dims <- length(dim(data)) # ndim of data
+  nvars <- length(var_names) #  vars names set by user
+
+  if(has_T_timeAxis) stopifnot('time' %in% nc_dims)
+  if(has_Z_verticalAxis) stopifnot('depth' %in% nc_dims)
+  if(has_T_timeAxis & has_Z_verticalAxis) stopifnot(data_dims == 3) # if have both T and Z, data should be 3 dims
+  if(!has_T_timeAxis & !has_Z_verticalAxis) stopifnot(nvars == nn) # if org by vars names should be equal to nl
+
+  if(verbose){
+    print(paste("The file has", nc$nvars, "variables and",  nc$ndim,"dimensions"))
+    print(paste("The dimensions are", paste(nc_dims, collapse = ', ')))
+  }
+  
+  # ---------------------------------------------------------------------
+  #  Locations and spatial info  ----------------------------------------
+  # ---------------------------------------------------------------------
+  
+  crs <- ncdf4::ncatt_get(nc, varid = 'crs')[['proj']]
+  
+  if(!inherits(locations, "Spatial") & is.null(crs)){
+    stop('Need CRS for locations data') # need crs if locations isn't spatially defined
+  }
+  
+  if(isGridded) {
+    if(inherits(locations, 'sf')) {
+      loc <- sf::st_geometry(locations)
+      st_crs(loc) <- crs 
+      loc <- as(loc, Class = 'Spatial')
+    } else { # either assumes its spatial class or data.frame
+      loc <- sp::SpatialPoints(locations)
```

Comment:
lines 907-908: simplify to `loc <- sp::SpatialPoints(locations, proj4string = sp::CRS(crs))`